Engine/Module/ModuleDebugDraw.cpp

```diff
@@ -0,0 +1,619 @@
+    {
+        assert(width > 0 && height > 0);
+        assert(pixels != nullptr);
+
+        GLuint textureId = 0;
+        glGenTextures(1, &textureId);
+        glBindTexture(GL_TEXTURE_2D, textureId);
+
+        glPixelStorei(GL_PACK_ALIGNMENT, 1);
+        glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+
+        glTexImage2D(GL_TEXTURE_2D, 0, GL_R8, width, height, 0, GL_RED, GL_UNSIGNED_BYTE, pixels);
+
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+
+        glBindTexture(GL_TEXTURE_2D, 0);
+        checkGLError(__FILE__, __LINE__);
+
+        return GLToHandle(textureId);
+    }
+
+    void destroyGlyphTexture(dd::GlyphTextureHandle glyphTex) override
+    {
+        if (glyphTex == nullptr)
+        {
+            return;
+        }
+
+        const GLuint textureId = handleToGL(glyphTex);
+        glBindTexture(GL_TEXTURE_2D, 0);
+        glDeleteTextures(1, &textureId);
+    }
+
+    // These two can also be implemented to perform GL render
+    // state setup/cleanup, but we don't use them in this sample.
+    //void beginDraw() override { }
+    //void endDraw()   override { }
+
+    //
+    // Local methods:
+    //
+
+    IDebugDrawOpenGLImplementation()
+        : mvpMatrix()
+        , width(0)
+        , height(0)
+        , linePointProgram(0)
+        , linePointProgram_MvpMatrixLocation(-1)
+        , textProgram(0)
+        , textProgram_GlyphTextureLocation(-1)
+        , textProgram_ScreenDimensions(-1)
+        , linePointVAO(0)
+        , linePointVBO(0)
+        , textVAO(0)
+        , textVBO(0)
+    {
+        //std::printf("\n");
+        //std::printf("GL_VENDOR    : %s\n",   glGetString(GL_VENDOR));
+        //std::printf("GL_RENDERER  : %s\n",   glGetString(GL_RENDERER));
+        //std::printf("GL_VERSION   : %s\n",   glGetString(GL_VERSION));
+        //std::printf("GLSL_VERSION : %s\n\n", glGetString(GL_SHADING_LANGUAGE_VERSION));
+        //std::printf("IDebugDrawOpenGLImplementation initializing ...\n");
+
+        // Default OpenGL states:
+        glEnable(GL_CULL_FACE);
+        glEnable(GL_DEPTH_TEST);
+        glDisable(GL_BLEND);
+
+        // This has to be enabled since the point drawing shader will use gl_PointSize.
+        glEnable(GL_PROGRAM_POINT_SIZE);
+
+        setupShaderPrograms();
+        setupVertexBuffers();
+
+        //std::printf("IDebugDrawOpenGLImplementation ready!\n\n");
+    }
+
+    ~IDebugDrawOpenGLImplementation()
+    {
+        glDeleteProgram(linePointProgram);
+        glDeleteProgram(textProgram);
+
+        glDeleteVertexArrays(1, &linePointVAO);
+        glDeleteBuffers(1, &linePointVBO);
+
+        glDeleteVertexArrays(1, &textVAO);
+        glDeleteBuffers(1, &textVBO);
+    }
+
+    void setupShaderPrograms()
```

Comment:
Could This code about shaders be replace by our ModuleProgram?

Reply:
Yes, I will work on that